### PR TITLE
Musan mixing implementation for multilingual japanese english recipe

### PR DESCRIPTION
**Pull request by baileyeet**

### CODE REVIEW

File: egs/multi_ja_en/ASR/local/utils/asr_datamodule.py [hunk 1]
```diff
@@ -15,7 +15,7 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-
+import torch
```
Comment (yuta0306) [inline egs/multi_ja_en/ASR/local/utils/asr_datamodule.py:18]: Move it to the second import block. You should use isort.

File: egs/multi_ja_en/ASR/local/utils/asr_datamodule.py [hunk 2]
```diff
@@ -265,12 +295,16 @@ def train_dataloaders(
             logging.info("Loading sampler state dict")
             train_sampler.load_state_dict(sampler_state_dict)
 
+        seed = torch.randint(0, 100000, ()).item()
```
Comment (yuta0306) [inline egs/multi_ja_en/ASR/local/utils/asr_datamodule.py:298]: Is this seed a fixed value? If not, I think it is better to fix seeds before using random functions.

File: egs/multi_ja_en/ASR/local/utils/asr_datamodule.py [hunk 3]
```diff
@@ -265,12 +295,16 @@ def train_dataloaders(
...
+        worker_init_fn = _SeedWorkers(seed)
```
Comment (yuta0306) [inline egs/multi_ja_en/ASR/local/utils/asr_datamodule.py:299]: Perfect!! 💯

File: egs/multi_ja_en/ASR/local/utils/update_cutset_paths.py [hunk 4]
```diff
@@ -19,44 +19,34 @@ def update_paths(cuts: CutSet, dataset_name: str, old_feature_prefix: str = "dat
         old_feature_prefix: The prefix that the original feature paths were relative to.
                             This typically corresponds to the root of the manifests dir
                             in the original recipe.
-    """
+"""
```
Comment (yuta0306) [inline egs/multi_ja_en/ASR/local/utils/update_cutset_paths.py:22]:
```suggestion
    """
```

File: egs/multi_ja_en/ASR/local/utils/update_cutset_paths.py [hunk 5]
```diff
@@ -19,44 +19,34 @@ def update_paths(cuts: CutSet, dataset_name: str, old_feature_prefix: str = "dat
...
-            # We assume old_feature_prefix is 'data/manifests'
-            # and original_storage_path looks like 'data/manifests/feats_train/feats-12.lca'
-            # We want to change it to 'data/manifests/<dataset_name>/feats_train/feats-12.lca'
+            try:
+                relative_path = original_storage_path.relative_to(old_feature_prefix)
+            except ValueError:
+				# If for some reason the path doesn't start with old_feature_prefix,
+				# keep it as is. This can happen if some paths are already absolute or different.
+				logger.warning(f"Feature path '{original_storage_path}' does not start with '{old_feature_prefix}'. Skipping update for this cut.")                
+                updated_cuts.append(cut)
```
Comment (yuta0306) [inline egs/multi_ja_en/ASR/local/utils/update_cutset_paths.py:33]: Strange indent

File: egs/multi_ja_en/ASR/local/utils/update_cutset_paths.py [hunk 6]
```diff
@@ -19,44 +19,34 @@ def update_paths(cuts: CutSet, dataset_name: str, old_feature_prefix: str = "dat
...
+                continue
+
+            # Avoid double-nesting (e.g., reazonspeech/reazonspeech/...)
+            # Construct the new path: data/manifests/<dataset_name>/feats_train/feats-12.lca
+			if relative_path.parts[0] == dataset_name:
+                    new_storage_path = Path("data/manifests") / relative_path
```
Comment (yuta0306) [inline egs/multi_ja_en/ASR/local/utils/update_cutset_paths.py:39]: Strange indent

File: egs/multi_ja_en/ASR/local/utils/update_cutset_paths.py [hunk 7]
```diff
@@ -77,6 +67,28 @@ def update_paths(cuts: CutSet, dataset_name: str, old_feature_prefix: str = "dat
...
+    musan_manifest_path = multi_recipe_manifests_root / "musan" / "musan_cuts.jsonl.gz"
+    if musan_manifest_path.exists():
+        logger.info(f"Processing musan manifest: {musan_manifest_path}")
+        try:
+           musan_cuts = load_manifest(musan_manifest_path)
+           updated_musan_cuts = update_paths(
+                   musan_cuts,
+                   "musan",
+                   old_feature_prefix="data/fbank"
+                   )
```
Comment (yuta0306) [inline egs/multi_ja_en/ASR/local/utils/update_cutset_paths.py:79]:
```suggestion
           updated_musan_cuts = update_paths(
               musan_cuts,
               "musan",
               old_feature_prefix="data/fbank"
           )
```

File: egs/multi_ja_en/ASR/local/utils/asr_datamodule.py [hunk 8]
```diff
@@ -192,7 +199,28 @@ def train_dataloaders(
 
         transforms = []
         input_transforms = []
+        if self.args.enable_musan:
+            logging.info("Enable MUSAN")
+            logging.info("About to get Musan cuts")
+            cuts_musan = load_manifest(self.args.manifest_dir / "musan/musan_cuts.jsonl.gz")
+            transforms.append(CutMix(cuts=cuts_musan, p=0.5, snr=(10,20), preserve_id=True))
```
Comment (kinanmartin) [inline egs/multi_ja_en/ASR/local/utils/asr_datamodule.py:206]: Just making a comment to mark it: these are the parameters that we may want to adjust during training in order to have more control over the noise settings in case English performance turns out to be poorer than expected or the language selection bias is too large. See [here](https://github.com/lhotse-speech/lhotse/blob/fda1a986e5e1e72a14c82049b4ee709fc09a81e6/lhotse/dataset/cut_transforms/mix.py#L16-L45) for the meaning of those parameters. 

The parameters you've chosen here are the same as what we've seen in existing MUSAN implementations though, so I think this all good!

**Comment by kinanmartin** [comment]
Looking it over, it does seem this change implements noise mixing for the entire combined dataset (MLS English and ReazonSpeech) as opposed to noise mixing for only the portion that is MLS English data. After looking over the recipe, I think that in order to have control over exactly which sub dataset gets the noise mixing, the noise mixing might have to be implemented [in this file](https://github.com/reazon-research/icefall/blob/6baa4b74d08ca19b6d313ad52f2d517dff3b56eb/egs/multi_ja_en/ASR/zipformer/multi_dataset.py) where the cuts are selected. It also may be a bit harder to implement because we would like to use the dataloader's  "transform" mechanism to apply the noise mixing, but we don't want to have to make a smaller dataloader just to mix noise into the MLS English cuts, then combine those cuts, then use a second main dataloader for the actual training. 

However, I think that it is still useful to have this implemented at the level of the combined dataset, so I think we can go with this implementation now and revisit if it seems the model is struggling more with Japanese.

**Review by kinanmartin — APPROVED** [review]
